sklearn/pipeline.py

```diff
@@ -93,6 +94,17 @@ class Pipeline(_BaseComposition):
         must define `fit`. All non-last steps must also define `transform`. See
         :ref:`Combining Estimators <combining_estimators>` for more details.
 
+    transform_input : list of str, default=None
+        This enables transforming some input arguments to ``fit`` (other than ``X``)
+        to be transformed by the steps of the pipeline up to the step which requires
+        them. Requirement is defined via :ref:`metadata routing <metadata_routing>`.
+        This can be used to pass a validation set through the pipeline for instance.
+
+        See the example TBD for more details.
```

Comment:
I'm very keen to see that example, maybe the HGBT early stopping case?

Reply:
I went and checked if I could use lighgbm, but there the validation set is passed as a list of tuples. No way to process that in a pipeline.

As for HGBT, it would look like this:

```py
from sklearn.datasets import make_regression
from sklearn.ensemble import HistGradientBoostingRegressor
from sklearn.model_selection import GridSearchCV, ShuffleSplit
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler

X, y = make_regression(n_samples=200, n_features=500, n_informative=5, random_state=0)
X[:2,] = X[:2,] + 20

# Validation set chosen before looking at the data.
X_val, y_val = X[:50,], y[:50,]
X, y = X[50:,], y[50:,]

est_gs = GridSearchCV(
    Pipeline(
        (
            StandardScaler(),
            HistGradientBoostingRegressor(
                early_stopping=True,
                validation_fraction=ShuffleSplit(test_size=0.3, random_state=123),
            ).set_fit_request(X_val=True, y_val=True),
        ),
        # telling pipeline to transform these inputs up to the step which is
        # requesting them.
        transform_input=["X_val", "y_val"],
    ),
    param_grid={"histgradientboostingregressor__max_depth": list(range(5))},
    cv=5,
).fit(X, y, X_val, y_val)
# this passes X_val, y_val to Pipeline, and Pipeline knows how to deal with
# them.
```

Reply:
To confirm my understanding:

```
est_gs = GridSearchCV( Pipeline( ( StandardScaler(), HistGradientBoostingRegressor( early_stopping=True, ).set_fit_request(X_val=True, y_val=True), ), ), param_grid={"histgradientboostingregressor__max_depth": list(range(5))}, cv=5, ).fit(X, y, X_val, y_val)
```

This would only transform `X`, and `y`, whereas your example would now explicitly mark `X_val, y_val` (and `X` and `y`) to be transformed as well?

Reply:
Correct!

Reply:
Only side remark: Our `HistGradientBoostingRegressor` does not YET support `X_val` and `y_val` in `fit`.